Feature/grpc redo networking communication

LiamClark opened the pull request:
1. actually generate the protobuffer code instead of having it in the repository as done by the example app
2. use grpc instead of stupid routing with their own socket code.
3. introduce a repository interface for their storage+ an inmemory non android dependent implementation so we can use it in tests.
4. unit test block creation through our grpc protocol

LiamClark committed:
Initial setup for grpc

This removes the jar dependency in the project on protobuf.
The message.proto class is now actually generated at build time not always present
in the project.

also includes all grpc dependencies but they are not used yet.

LiamClark committed:
Grpc integration test, might be a bit dodgy but seems to work on android

LiamClark committed:
Throw away inheritance in communication

LiamClark committed:
Sketch of grpc communication

LiamClark committed:
In memory blockchain storage

LiamClark committed:
First integration test for grpc

To achieve this move the implementation classes in the grpc test
to actual production code.

Always insert a genesis block into the inmemory database.
Moved some more code to our db interface rather than the sql
implementaiton.

Also removed some last compiler error regarding bluethooth

LiamClark committed:
Save crawled blocks in grpc

In order to do this

1. remove the stupid base64 conversion in validation, to remove an
android api call.
2. make the methods in in memory storage return nullable things, since
this does happen.
3. reformat the Key file.
4. return a list instead of an iterator since after traversing it once
to save all the records the iterator is then empty.

LiamClark committed:
Allow for block creation with a transaction using grpc

This nearly completely finishes the port of their communication
to ours.

The test covers nearly every nice weather path in an integration
fashion.

Further testing for bad weather should be done with a rewrite of
validate cause fuck that method.

LiamClark committed:
Bind our gpc server in the main activity.

Manual test of creating a block on two devices worked!
we need to properly translate the port of the peer item.

it was wrong on 8081 that's the port used for the discovery,
not the port we use for the services.

LiamClark committed:
Regresion test for bytes to hex

GijsWeterings commented on app/build.gradle:
```diff
@@ -39,6 +75,15 @@ dependencies {
+
+    //grpc
+    // You need to build grpc-java to obtain these libraries below.
+//    compile 'io.grpc:grpc-okhttp:1.8.0'
```
Is this no longer needed?

Griezzly commented on app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainActivity.kt:
```diff
@@ -56,7 +57,9 @@ class MainActivity : AppCompatActivity(), CommunicationListener {
-        communication!!.connectToPeer(peer, payload)
+
+        server.connectToPeer(peeritem.withPort(8080))
+        server.sendBlockToKnownPeer(peeritem.withPort(8080), payload)
```
we should probably use a less commonly used port and not hardcode it like this

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/Util/Key.java:
```diff
@@ -184,15 +192,31 @@ public static PublicKey loadPublicKey(String key) {
+        try {
+            return kf.generatePublic(pubKeySpec);
+        } catch (InvalidKeySpecException e) {
+            e.printStackTrace();
```
can we log this properly? I understand you don't want to tie this to the android logger, but maybe log4j/logback is an option?

By the way, any reason this class is kept as Java?

LiamClark replied:
Yes, I just needed this overload to avoid an android dependency for base64 encoding.
they were taking bytes encodignt o base 64 and then back to bytes... 
and the base 64 was an android dependency. furthermore the exception was never properly logged in the other overload either. I left this as java since I wanted to modify this file as less as possible

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/attestation/Attestation.kt:
```diff
@@ -0,0 +1,44 @@
+
+class RangeProof() : ZeroKnowledgeProof()
+
+//sort of saved this as a block super cool
```
sort of?

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/attestation/Attestation.kt:
```diff
@@ -0,0 +1,44 @@
+}
+
+fun verifier(): Single<Boolean> {
+    val challenge = Pair(20, 50)
```
These are all arbitrary values as placeholder, right?

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/block/TrustChainBlock.java:
```diff
@@ -356,10 +354,9 @@ else if(nextBlock.getSequenceNumber() != block.getSequenceNumber() + 1) {
-     * @param db - Database to validate against
      * @return a VALID validation result
      */
-    public static ValidationResult validateTransaction(MessageProto.TrustChainBlock block, SQLiteDatabase db) {
```
The db is still being used, but is decoupled, correct?

LiamClark replied:
yes, but this never even used the database this is their place holder for validation, we won't be doing it in there we will do it where we think it's fit, this is just to make stuff compile easier

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/connection/Communication.kt:
```diff
@@ -1,47 +1,31 @@
 
 /**
- * Class that is responsible for the communication.
+ * Leaving this here so we can see what they were doing.
```
// FIXME remove?

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/connection/Communication.kt:
```diff
@@ -380,11 +333,34 @@ abstract class Communication(private val dbHelper: TrustChainDBHelper, private v
+    }
+
+    fun sendMessage(peer: Peer, message: MessageProto.Message) {
+//        val task = ClientTask(
```
you just nuked this to prevent it from interfering with the new way, I presume?

LiamClark replied:
that file is gone since it's completely replaced by grpc, I kept the communication to sort of glipmse at the original protocol but that can be removed completely now too

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+            val signBlock = signBlock(peer, block)
+            val returnTrustChainBlock = ChainService.PeerTrustChainBlock.newBuilder().setPeer(me).setBlock(signBlock).build()
+            responseObserver.onNext(returnTrustChainBlock)
+            responseObserver.onCompleted()
```
Do we never complete the response if a crawl has to be made, or is that somehow handled?

LiamClark replied:
not yet, how about we complete the stream with an error?

GijsWeterings replied:
Sounds good to me

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+
+
+    override fun getPublicKey(request: ChainService.Empty, responseObserver: StreamObserver<ChainService.Key>) {
+        responseObserver.onNext(createPrivateKey())
```
You gave me a heart attack here. `createPrivateKey` in `getPublicKey`'s response. Luckily, the `createPrivateKey` method returns the public key... You should probably rename that ;)

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+        try {
+            validation = TrustChainBlock.validate(block, dbHelper)
+        } catch (e: Exception) {
+            e.printStackTrace()
```
Again, we should probably log this

GijsWeterings replied:
Also, can we catch more specific Exceptions here?

LiamClark replied:
it doesn't actually throw any exceptions... 
they wrapped it around it I have no fucking clue why...

GijsWeterings replied:
![](https://media.giphy.com/media/3o7aCPyuFTbB9uuady/giphy.gif)

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+        // determine if we should sign the block, if not: do nothing
+        if (!shouldSign(block)) {
+            Log.e(TAG, "Will not sign received block.")
+            return null
```
Would it be a good idea to give a validationstatus == `wontsign` or something?

LiamClark replied:
I aim to refactor that later when we actually start adding our own business logic.

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+
+    fun connectToPeer(peer: PeerItem): List<ChainService.PeerTrustChainBlock> {
+        val connectablePeer = peer.asPeerMessage()
+        val crawledBlocks = sendCrawlRequest(connectablePeer, myPublicKey, -5)
```
So we always ask for the last 5 blocks? Is that enough?

LiamClark replied:
They seemed to think so

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+        val crawlResponse = ChainService.CrawlResponse.newBuilder().setPeer(me).addAllBlock(crawledBlocks).build()
+        val peerMessage = peer.asPeerMessage()
+        val peerChannel = registry.channelForPeer(peerMessage)
+        val theirPublicKey: ChainService.Key = peerChannel.sendLatestBlocks(crawlResponse)
```
sendLatestBlocks returns the other party's public key?

LiamClark replied:
Yup, a bit weird, this was before we had the getPublicKey message, we could remove this but it saves a message

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/grpc/ChainServiceServer.kt:
```diff
@@ -0,0 +1,280 @@
+
+    companion object {
+        fun shouldSign(block: MessageProto.TrustChainBlock): Boolean {
+            return true
```
Since we're going to have to make this part "interactive", should this be on the companion object?

LiamClark replied:
probably not, out of scope for this pr though

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainActivity.kt:
```diff
@@ -129,18 +127,45 @@ class MainActivity : AppCompatActivity(), CommunicationListener {
+        get() {
+            try {
+                val netInterfaces = NetworkInterface.getNetworkInterfaces()
+                for (netInt in netInterfaces) {
```
omg loops

LiamClark replied:
I decided to be sort of nice for others

GijsWeterings commented on app/src/main/java/nl/tudelft/cs4160/identitychain/main/MainActivity.kt:
```diff
@@ -129,18 +127,45 @@ class MainActivity : AppCompatActivity(), CommunicationListener {
         //start listening for messages
-        communication!!.start()
+        requireNotNull(localIPAddress, { "error could not find local IP" })
+        val (server, grpc) = ChainServiceServer.createServer(kp, 8080, localIPAddress!!, dbHelper)
```
Are we just losing the grpc reference here?

LiamClark replied:
yeah, we currently just leak the resource, it's only needed for shutting the server down when the app exits, let's make an issue for that

GijsWeterings commented on app/src/test/java/nl/tudelft/cs4160/identitychain/PeerTest.java:
```diff
@@ -0,0 +1,18 @@
+
+import static org.junit.Assert.*;
+
+public class PeerTest {
```
No longer needed, I presume?

LiamClark replied:
yep

LiamClark committed:
Add rxgrpc

LiamClark committed:
Fix some feedback.

deleted communication and peer test.
Reply to recieve half block with an error
if the chain has gaps.

and rename public key -> private key

GijsWeterings commented:
Unhandled feedback are standing problems, so we can merge
